design/one-pager-moving-packages.md

```diff
@@ -0,0 +1,462 @@
+package metadata, and are fully qualified - they specifically include the OCI
+registry to pull from. You can replicate a package from `xpkg.upbound.io` to
+`pkg.internal.acme.co`, but its dependencies are still hard coded to
+`xpkg.upbound.io`.
+
+## Goals
+
+The goal of this proposal is to support:
+
+1. Moving a dependency to a different registry, organization, or repository
+2. Replicating a dependency to a different, potentially private, registry
+
+Handling private registry authentication for dependencies isn't in scope for
+this design. It's covered by the [package `ImageConfig` proposal][imageconfig].
+
+## Proposal
+
+I propose we do two things:
+
+1. Allow packages to specify that they replace another package
+2. Allow packages to specify partially qualified dependencies
+
+These proposed changes correspond to the goals above. The first addresses
+moving dependencies, while the second addresses replicating dependencies.
+
+### Moving a Package
+
+I propose we support moving a package by adding a `replaces` stanza to package
+metadata.
+
+For example, assume the maintainers of `provider-family-aws` move it from
+`xpkg.upbound.io/upbound/provider-family-aws` to
+`xpkg.upbound.io/crossplane-contrib/provider-family-aws`. The maintainers would
+add the following `replaces` stanza to the moved provider:
+
+```yaml
+apiVersion: meta.pkg.crossplane.io/v1
+kind: Provider
+metadata:
+  name: provider-family-aws
+spec:
+  replaces:
+  - xpkg.upbound.io/upbound/provider-family-aws
+```
+
+This tells Crossplane that the `crossplane-contrib/provider-family-aws` package
+replaces the `upbound/provider-family-aws` package.
+
+When the revision controller encounters a package with a replaces stanza, it'll
+deactivate the replaced package(s) before installing the new package.
+
+When the replaced package is deactivated, the new package can be installed
+without a conflict. A deactivated package [owns][owner-ref] but doesn't
+[control][controller-ref] its payload (e.g. a Provider's MR CRDs). This
+eliminates the controller reference conflict.
+
+A deactivated package exists, but in an inactive state. Specifically:
+
+* It appears under (e.g.) `kubectl get providers`
+* At least one revision appears under (e.g.) `kubectl get providerrevisions`
+* The package's `spec.revisionActivationPolicy` is set to `Manual`
+* All revisions have `spec.desiredState` set to `Inactive`
+* It has no runtime - e.g. no Deployment of a provider's runtime binary
+* It owns but doesn't control its package payload - e.g. a provider's MR CRDs
+
+Crucially, deactivating a package doesn't remove its payload. For example
+deactivating a provider doesn't remove its MR CRDs like uninstalling it would.
+The new package shares ownership (in the [owner reference][owner-ref] sense)
+with the replaced package, until the replaced package is manually uninstalled.
+This allows the package manager to seamlessly handover ownership and control of
+a package's payload.
```

Comment:
I assume one can safely delete a deactivated/replaced package at this moment.

Are there any good reasons not to automate this as well, other than the possible complexity of differentiating a replaced and deactivated package from another explicitly deactivated package?

Reply:
One possible drawback just came to my mind is that, if a package mistakenly (or maliciously) replaces a package that it actually should not, it may end up with uninstallation of manifests (e.g. CRDs) of the original package.

Let's say I've added a replace stanza in `provider-foo` stating that it replaces `xpkg.upbound.io/upbound/provider-aws-rds`. If we automate deletion as well, I'll probably get my RDS CRDs deleted (yes, it will stuck and not delete the actual external resources since its controller not running/deactivated, but still a mess).

Reply:
Yeah, I think the safest/most conservative thing would be to deactivate but not delete the replaced package.

In a way you could think about it as the replacement package being a new revision of the replaced package. If you think about it that way it makes sense to deactivate the replaced package's revisions but keep them around. Just like we'd do for a new revision of a (regular, not replaced) package.

Even if a replacement package isn't malicious, it's possible it drops a type (e.g. an XRD or an MR CRD) that the replaced package had installed. I think it's safest to not automatically delete that type.